### PR TITLE
Easier integration of the Control Toolbox into build systems

### PR DESCRIPTION
**Pull request by schulz0r**
Dear developer team,
I am using your Control Toolbox for path planning problems. Specifically, I include the toolbox as an external CMake project into my path planning software.
Usually, there should be a CMakeLists.txt in the root directory of a CMake project, but there is none, which is inconvenient. Also, there are a lot of sh-scripts in the `ct/` directory which perform basic tasks like installing or building the code. These can be done by CMake itself, so I don't understand what's the prupose of these scripts.

I wanted to conveniently integrate the toolbox into my build system, so I created the fork which restructured the project to have a root CMakeList. This list adds the modules like `optcon` etc. as submodules.
Besides this, I changed the name of the project from _CT_ to _ControlToolbox_ and `optcon` to OptimalControl etc.. So, after using `find_package(ControlToolbox REQUIRED)` in my project, I can add its submodules by linking my target against `ControlToolbox`, `ControlToolbox::OptimalControl` and so on.
I renamed all headers which did not have an extension to end with .h, because there was a git bug which prevented these files from being cloned onto my disk.
Another thing is that I think that documentation, examples and predefined models should not be build by default as it slows down the automated build process while wasting disk space. Thats why I added options checking to the root CMakeLists, which check if `-DBUILD_EXAMPLES`, `-DBUILD_DOCS` and `-DBUILD_MODELS` are set.
The last thing I changed was that when QT is not found, then examples using a GUI are not being built as there will be a compile error anyway.

There is still alot to do to get rid of the scripts. For instance, it is possible to add dependencies as an external project in case `find_package()` could not find them. This will work on any machine, whereas your script uses `apt` which is only available on some linux systems. The best way to do this is described in the google test [readme](https://github.com/google/googletest/blob/master/googletest/README.md).

So, I hope that you can make use of my changes. I have to warn you thou, because I have no overview about all features you implemented in this project so I have not tested everything, for example if ROS is still working. I would be glad to hear about your opinion.

### CODE REVIEW

**Comment by markusgft** [comment]
Dear @schulz0r 
First of all thanks a lot for your interest in contributing to this open-source project, new improvements are always very welcome! 
The pullrequest as it is presented here is quite extensive (more than 200 changed files). In the interest of helping us to provide high-quality reviews and feedback, do you think there would be the possiblity to transform it into smaller, self-contained, more reviewable pieces? Also, you should know that there is an active development effort going on in the background -- merging such large PRQ into the development repo will be very hard.

> Usually, there should be a CMakeLists.txt in the root directory of a CMake project, but there is none, which is inconvenient. Also, there are a lot of sh-scripts in the `ct/` directory which perform basic tasks like installing or building the code. These can be done by CMake itself, so I don't understand what's the prupose of these scripts.

Le me quickly comment on this. I agree that the default use of cmake assumes a structure where we have a single project. I have been considering combining all four CT domains (core, optcon, rbd and models) for quite a while. The main challenge is that we have a quite diverse user-base, of which roughly 50% are not interested in using (and therefore building and installing) the robotics features, i.e. ct_rbd and ct_models. If we would combine everything into one single project, what would be the most elegant way of ensuring that people not interested in ct_rbd and ct_models do not have to install them, alongside their additional dependencies such as _kindr_?

> I wanted to conveniently integrate the toolbox into my build system, so I created the fork which restructured the project to have a root CMakeList. This list adds the modules like `optcon` etc. as submodules.

Would there be a way of keeping ct_core as a submodule, too? I still like the idea of having the code of core, optcon, etc. at equivalent hierarchical levels. The dependency graph is
```core -- optcon -- rbd -- models```
but what the new structure displays is
```
core
optcon -- rbd --models
```

> Besides this, I changed the name of the project from _CT_ to _ControlToolbox_ and `optcon` to OptimalControl etc.. So, after using `find_package(ControlToolbox REQUIRED)` in my project, I can add its submodules by linking my target against `ControlToolbox`, `ControlToolbox::OptimalControl` and so on.

Why is this required? This is a feature which is not necessary to achieve compatibility and may even break existing code from our users when upgrading. From our perspective, we would prefer to stick with the old names.

> I renamed all headers which did not have an extension to end with .h, because there was a git bug which prevented these files from being cloned onto my disk.

First of all, could we verify and document that the bug that you are mentioning is affecting all potential Mac users? If this is in broad-scale issue for OSX, we will need to change something in the CT, otherwise a local workaround should be sufficient. Nevertheless, having these collective (meta) headers like `State`, `Common` simply renamed to the same name plus a `*.h` extension would not be my first preference. Maybe we should think of a more expressive renaming in this case?

> Another thing is that I think that documentation, examples and predefined models should not be build by default as it slows down the automated build process while wasting disk space.

They (documentation and examples) have never been build by default, this has always been optional on v3.0.2.

 Thats why I added options checking to the root CMakeLists, which check if `-DBUILD_EXAMPLES`, `-DBUILD_DOCS` and `-DBUILD_MODELS` are set.
> The last thing I changed was that when QT is not found, then examples using a GUI are not being built as there will be a compile error anyway.

Good point.

> There is still alot to do to get rid of the scripts. For instance, it is possible to add dependencies as an external project in case `find_package()` could not find them. This will work on any machine, whereas your script uses `apt` which is only available on some linux systems. The best way to do this is described in the google test [readme](https://github.com/google/googletest/blob/master/googletest/README.md).

Quick question at this point, since I have not used this mechanism before: so it would be possible to only have the external project pulled and installed whenever find_package() returns false? I think, however, we will need some more detailed solution here. For example, we have quite a number of users who are not  allowed to install some of the dependencies, like `CppAD-Codgen`, `hpipm` or `ipopt`, for license reasons. So you see, there is a number of dependencies which should not simply be pulled automatically if find_package() is returns false.

> So, I hope that you can make use of my changes. I have to warn you thou, because I have no overview about all features you implemented in this project so I have not tested everything, for example if ROS is still working. I would be glad to hear about your opinion.

Wrapping up, I am very interested in supporting you in introducing a cmake-integration which works for your operating system. However, we need to maintain strict compatibility with users of standard linux distributions like Ubuntu and Mint, as well as ROS build system users. I just pulled your fork to my Ubuntu 16.04 machine, cmake fails after a few seconds reporting
`fatal error: Eigen/Dense: No such file or directory`. Therefore, I think in some way or the other, we need to enable you to test your changes on Ubuntu and with `catkin`. Would you be willing to use a virtual machine with Ubuntu/ROS for testing your developments?

Thanks again, I do think that you are up to making a really cool contribution to the CT.  I hope that you are interested in investing more time to resolve some of the mentioned issues, it will certainly be appreciated by many users.

--- 
Edit: it might make sense to also include @acxz in this discussion, since he maintains the arch-linux package of the CT and might have additional thoughts and requirements.

**Comment by acxz** [comment]
@schulz0r This is a lot of work here in your PR. There are definitely some useful things here and I actually understand some of your concerns as the Arch Linux maintainer of this codebase, (having to work around the shell scripts and nonstandard CMake structure)

As @markusgft has mentioned I believe the first course of action to do here would be to try to clean this PR up by breaking it into different pieces that target a specific issue/concern. I would create a issue for each issue/enhancement you have, and then create a PR with the commits that directly address that. This creates a trace down and a platform for proper discussion so that your changes can be easily reviewed/incorporated into the main source tree if deemed fit. This is a lot of work to digest and maintaining compatibility is important.

Many of the changes specified here will definitely be breaking and I would have to rewrite some of the build/package recipes for Arch Linux, although that shouldn't be a hassle. But it should only be done on a minor version change (if we are following semantic versioning).

Maybe you can get some useful feedback by looking at how I build and package control-toolbox for Arch Linux.
Essentially each `core`, `opt_con`, etc. is a separate package and there is one meta package that pulls all of them in. [Here](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=control-toolbox-optcon) is the build scripts for one of them, you can find more by searching the AUR. I don't utilize the shell scripts at all, but rather just the individual CMakelists.txt in each project, and that works completely fine.

Personally I think that these changes are too drastic to just get control-toolbox to work with your build system and I feel that adapting your build system a little bit to the structure of control-toolbox might have been a better idea.
However there are some useful things here and the discussion sprouted from this *is* meaningful. I think let's first open targeted issues for each aspect of this PR and break this PR down to address each issue you open up. This will provide us a digestible format and a systematic way to move forward.

As always contributions are most welcome and thx for the work you have done, we just need to make sure one person's code doesn't break another one's, haha.

Also @markusgft 
> what would be the most elegant way of ensuring that people not interested in ct_rbd and ct_models do not have to install them, alongside their additional dependencies such as kindr?

We could accomplish this with build flags inside of the root CMakeLists.txt

**Comment by schulz0r** [comment]
Okay, that looks like there is still some work to do. I would invest some time to get the rework done, so we should start with easily agreeable changes like the missing .h extensions or bugs. I have experience with ROS and CI Pipelines (basics) so I know how to work with CMake in that regard.
> If we would combine everything into one single project, what would be the most elegant way of ensuring that people not interested in ct_rbd and ct_models do not have to install them, alongside their additional dependencies such as kindr?

I think acxz already posted a good solution. Other than that, I did not combine everything in one single project, so I don't know if I get you right.

> The dependency graph is `core -- optcon -- rbd -- models` but what the new structure displays is
```
core
optcon -- rbd --models
```

That's supposed to work better with CMake. CMake works with a top-down hierarchy and core is on the top. I has to be compiled no matter which additional submodules are desired. The core target is then passed down the hierarchy to optional modules. Before, all modules were on the same hierarchy level so you would have to create interdependencies between modules. CMake can do this, but why putting in the effort?

> A *.h extension would not be my first preference.

Why thou? This is what you expect a header to look like when working with C/C++.

> Quick question at this point, since I have not used this mechanism before: so it would be possible to only have the external project pulled and installed whenever find_package() returns false?

You don't need to install external projects. I never do this because it requires privileged access.
You can simply pull the external project and compile it during configuration or simply add it to the project. Also, packages which need a license should not automatically be added to the project. From what I've seen, it is possible to compile the library without dependencies like `CppAD-Codgen`. For the other dependencies which provide the basic functionality of the toolbox, see this excerpt of one of my CMake files:
```
find_package(Eigen3 3.3.9 QUIET NO_MODULE)

if((NOT ${Eigen3_FOUND}) AND (NOT TARGET Eigen3-download))
    # Download and unpack eigen3 at configure time
	configure_file(CMakeLists-Eigen3.txt.in Eigen3-download/CMakeLists.txt)
	execute_process(COMMAND ${CMAKE_COMMAND} -G "${CMAKE_GENERATOR}" -DBUILD_TESTING=OFF .
	  RESULT_VARIABLE result
	  WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/Eigen3-download )
	if(result)
	  message(FATAL_ERROR "CMake step for Eigen3 failed: ${result}")
	endif()
	execute_process(COMMAND ${CMAKE_COMMAND} --build .
	  RESULT_VARIABLE result
	  WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/Eigen3-download )
	if(result)
	  message(FATAL_ERROR "Build step for Eigen3 in commandmodel failed: ${result}")
	endif()
	
	add_subdirectory(${CMAKE_CURRENT_BINARY_DIR}/Eigen3-src
	${CMAKE_CURRENT_BINARY_DIR}/Eigen3-build
	  EXCLUDE_FROM_ALL)
endif()

if(TARGET Eigen3)
    add_dependencies(${PROJECT_NAME} Eigen3_download)
endif()

target_link_libraries(${PROJECT_NAME} PRIVATE Eigen3::Eigen)
```
CMakeLists-Eigen3.txt.in looks like this:
```
cmake_minimum_required(VERSION 3.10)

project(Eigen3-download NONE)

include(ExternalProject)
ExternalProject_Add(Eigen3
    GIT_REPOSITORY    https://gitlab.com/libeigen/eigen.git
    SOURCE_DIR        "${CMAKE_CURRENT_BINARY_DIR}/Eigen3-src"
    BINARY_DIR	      "${CMAKE_CURRENT_BINARY_DIR}/Eigen3-build"
    BUILD_COMMAND     ""
    GIT_PROGRESS      On
    UPDATE_COMMAND    ""
    CONFIGURE_COMMAND ""
    INSTALL_COMMAND   ""
    TEST_COMMAND      ""
)
```
That Eigen3 was missing at your first attempt to compile this is because I did not implement above code yet, I just mentioned that it is possible to get rid of the dependency installation script.

@acxz I don't know about packaging for arch linux, but I usually use CPack to create linux packages, which also has a CMake integration. That's how packaging can be automated very easily. Creating packages for each submodule should be easily done with a cpack command for each submodule as soon as the CMakeList contains the package information. I can look into that if you want.

In conclusion, I think all problems which have been addressed here can be solved with medium effort, so let me hear where to start.

**Comment by acxz** [comment]
> CPack to create linux packages

I'm not that familiar with CPack but I don't think it works well with Arch Linux packages so that won't be a solution for me.



> let me hear where to start

From our comments:

> do you think there would be the possiblity to transform it into smaller, self-contained, more reviewable pieces?

and

> create a issue for each issue/enhancement you have, and then create a PR with the commits that directly address that

Please start with those suggestions

**Comment by schulz0r** [comment]
Okay, the PR will be closed and I will try again.
